gadget/update.go

```diff
@@ -159,52 +159,77 @@ func EnsureLayoutCompatibility(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVo
-			return false, "filesystems do not match"
-		case !IsCreatableAtInstall(gv) && dv.Filesystem != gv.Filesystem:
-			return false, "filesystems do not match and the partition is not creatable at install"
+		}
+
+		switch {
+		// on disk size too small
 		case dv.Size < gv.Size:
 			return false, fmt.Sprintf("on disk size %d (%s) is smaller than gadget size %d (%s)",
 				dv.Size, dv.Size.IECString(), gv.Size, gv.Size.IECString())
-		case gv.Role != SystemData && dv.Size > gv.Size:
-			return false, fmt.Sprintf("on disk size %d (%s) is larger than gadget size %d (%s) (and the role should not be expanded)",
-				dv.Size, dv.Size.IECString(), gv.Size, gv.Size.IECString())
-		default:
-			return false, "some other logic condition (should be impossible?)"
+
+		// on disk size too large
+		case dv.Size > gv.Size:
+			// larger on disk size is allowed specifically only for system-data
+			if gv.Role != SystemData {
+				return false, fmt.Sprintf("on disk size %d (%s) is larger than gadget size %d (%s) (and the role should not be expanded)",
+					dv.Size, dv.Size.IECString(), gv.Size, gv.Size.IECString())
+			}
+		}
+
+		// If we got to this point, the structure on disk has the same name,
+		// size and offset, so the last thing to check is that the filesystem
+		// matches (or that we don't care about the filesystem).
+
+		// TODO: here we need to handle in the strict case partitions which are
+		// to be created at install and are encrypted like ubuntu-data as they
+		// will not match filesystems exactly and need some massaging
+
+		if opts.AssumeCreatablePartitionsCreated || !IsCreatableAtInstall(gv) {
+			// we assume that this partition has already been created
+			// successfully - either because this function was forced to(as is
+			// the case when doing gadget asset updates), or because this
+			// structure is not created during install
+
+			// note that we only check the filesystem if the gadget specified a
+			// filesystem, this is to allow cases where a structure in the
+			// gadget has a image, but does not specify the filesystem because
+			// it is some binary blob from a hardware vendor for non-Linux
+			// components on the device that _just so happen_ to also have a
+			// filesystem when the image is deployed to a partition. In this
+			// case we don't care about the filesystem at all because snapd does
+			// not touch it, unless a gadget asset update says to update that
+			// image file with a new binary image file.
+			if gv.Filesystem != "" {
+				// then the gadget specified a filesystem and the on disk needs
+				// to match
+				if gv.Filesystem != dv.Filesystem {
+					// use more specific error message for structures that are
+					// not creatable at install
+					if !IsCreatableAtInstall(gv) {
+						return false, "filesystems do not match and the partition is not creatable at install"
+					}
+					// otherwise generic
+					return false, "filesystems do not match"
```

Comment:
It may help diagnosing these errors if you include the filesystem in the message
```suggestion
					fsToMsg := func(fs string) string { if fs != "" { return fs } else { return "(none)" } }
					errMsg := fmt.Sprintf("filesystems do not match (expected %q got %q)", fsToMsg(gv.Filesystem), fsToMsg(dv.Filesystem))
					if !IsCreatableAtInstall(gv) {
						return false, errMsg + " and the partition is not creatable at install"
					}
					// otherwise generic
					return false, errMsg
```

Reply:
good idea, done